Automated Market Maker (XLS-30d) reference documentation (#1541)
XRPLF/xrpl-dev-portal: content/references/protocol-reference/transactions/transaction-types/ammdeposit.md

---
html: ammdeposit.html
parent: transaction-types.html
blurb: Deposit funds into an Automated Market Maker in exchange for LPTokens.
labels:
  - AMM
status: not_enabled
---
# AMMDeposit
[[Source]](https://github.com/gregtatcam/rippled/blob/amm-core-functionality/src/ripple/app/tx/impl/AMMDeposit.cpp "Source")
<!-- TODO: Update source link to merged version when available -->

{% include '_snippets/amm-disclaimer.md' %}

Deposit funds into an Automated Market-Maker (AMM) instance and receive the AMM's liquidity provider tokens (_LP Tokens_) in exchange. You can deposit one or both of the assets in the AMM's pool.

## Example {{currentpage.name}} JSON

```json
{
  "TransactionType": "AMMDeposit",

>   "AMMID": "4B82E4CB90FBE26EA172A498F7A7B03A2A7C285149A9B18731370F2644B96F7A",

Reviewer: `AMMID` is replaced with `Asset/Asset2`.

  "Account": "rJVUeRqDFNs2xqA7ncVE6ZoAhPUoaJJSQm",

>   "Asset1In": {

Reviewer: Changed to `Amount/Amount2`.

    "issuer": "rP9jPyP5kyvFRb6ZiRghAGw5u8SGAmU4bd",
    "value": "2.5"
  },
  "Asset2In": "30000000",
  "Flags": 0,
  "Sequence": 4,
  "Fee": "12",
  "LastLedgerSequence": 369823
}
```

{% include '_snippets/tx-fields-intro.md' %}

| Field         | JSON Type           | [Internal Type][] | Required? | Description |
|:--------------|:--------------------|:------------------|:----------|:------------|
| `AMMID`       | String              | Hash256           | Yes       | The AMMID of the AMM instance to deposit into. **Caution:** This is not the ID of the AMM ledger object! |
| `Asset1In`    | [Currency Amount][] | Amount            | No        | The amount of one asset to deposit to the AMM. If present, this must match the type of one of the assets (tokens or XRP) in the AMM's pool. |
| `Asset2In`    | [Currency Amount][] | Amount            | No        | The amount of another asset to add to the AMM. If present, this must match the type of the other asset in the AMM's pool and cannot be the same asset as `Asset1In`. |
| `EPrice`      | [Currency Amount][] | Amount            | No        | The maximum effective price, in the deposit asset, to pay for each LP Token received. |

> | `LPTokensOut` | [Currency Amount][] | Amount            | No        | How many of the AMM's LP Tokens to buy. |

Reviewer: Changed to `LPTokenOut`.

### AMMDeposit Modes

This transaction has several modes, depending on which combination of fields are provided. The modes fall into two categories: 

- **Double-asset deposits**, in which you provide both assets in the AMM's pool in proportions that match the balance of those assets already there. These deposits are not subject to a fee.
- **Single-asset deposits**, in which you provide only one of the AMM's two assets. You can envision a single-asset deposit as trading _some_ of the deposited asset for the other asset, then performing a double-asset deposit. The trading fee applies to the amount you would need to trade for, but not to the rest of the deposit; the amount of the fee is debited from the LP Tokens paid out. _For example, if the AMM's asset pool is split perfectly evenly between USD and EUR, and you try to deposit 100 USD, the amount of LP Tokens you receive is slightly less than if you had deposited 50 EUR + 50 USD, because you pay the trading fee to convert some of your USD to an equal amount of EUR._

The following combinations of fields indicate a **double-asset deposit**:

> 

Reviewer: Flags are required (added below) to specify a valid set of parameters.

| Fields Specified          | Meaning |
|---------------------------|---------|

> | `LPTokensOut` only        | Deposit both of this AMM's assets, in amounts calculated so that you receive the specified amount of LP Tokens in return. The amounts deposited maintain the relative proportions of the two assets the AMM already holds. |

Reviewer: Flag `tfLPToken`: 0x01

> | `Asset1In` and `Asset2In` | Deposit both of this AMM's assets, up to the specified amounts. The actual amounts deposited must maintain the same balance of assets as the AMM already holds, so the amount of either one deposited MAY be less than specified. The amount of LP Tokens you get in return is based on the total value deposited. |

Reviewer: Flag `tfTwoAsset`:0x04

The following combinations of fields indicate a **single asset deposit**:

| Fields Specified             | Meaning |
|------------------------------|---------|

> | `Asset1In` only              | Deposit exactly the specified amount of one asset, and receive an amount of LP Tokens based on the resulting share of the pool (minus fees). |

Reviewer: Flag `tfSingleAssset`: 0x02

> | `Asset1In` and `LPTokensOut` | Deposit up to the specified amount of one asset, so that you receive exactly the specified amount of LP Tokens in return (after fees). |

Reviewer: Flag `tfOneAssetLPToken`: 0x08

> | `Asset1In` and `EPrice`      | Deposit up to the specified amount of one asset, but pay no more than the specified effective price per LP Token (after fees). |

Reviewer: Flag `tfLimitLPToken`:0x10

---
html: ammdeposit.html
parent: transaction-types.html
blurb: Deposit funds into an Automated Market Maker in exchange for LPTokens.
labels:
  - AMM
status: not_enabled
---
# AMMDeposit
[[Source]](https://github.com/gregtatcam/rippled/blob/amm-core-functionality/src/ripple/app/tx/impl/AMMDeposit.cpp "Source")
<!-- TODO: Update source link to merged version when available -->

{% include '_snippets/amm-disclaimer.md' %}

> Deposit funds into an Automated Market-Maker (AMM) instance and receive the AMM's liquidity provider tokens (_LP Tokens_) in exchange. You can deposit one or both of the assets in the AMM's pool.

Reviewer 2: ```suggestion
Deposit funds into an Automated Market-Maker (AMM) instance and receive the AMM's liquidity provider tokens (_LP Tokens_) in exchange. You can deposit either asset in the AMM's pool.
```

Author: The intended meaning is you can deposit A, B, or A+B (and the semantics of A+B are different than depositing just one). Suggested wording sounds like the options are A or B (single-asset deposit) only.

Reviewer 2: Misread the original. Keep as is.

| `Asset`       | Object              | STIssue           | Yes       | The definition for one of the assets in the AMM's pool. In JSON, this is an object with `currency` and `issuer` fields (omit `issuer` for XRP). |
| `Asset2`      | Object              | STIssue           | Yes       | The definition for the other asset in the AMM's pool. In JSON, this is an object with `currency` and `issuer` fields (omit `issuer` for XRP). |
| `Amount`      | [Currency Amount][] | Amount            | No        | The amount of one asset to deposit to the AMM. If present, this must match the type of one of the assets (tokens or XRP) in the AMM's pool. |
| `Amount2`     | [Currency Amount][] | Amount            | No        | The amount of another asset to add to the AMM. If present, this must match the type of the other asset in the AMM's pool and cannot be the same asset as `Amount`. |
| `EPrice`      | [Currency Amount][] | Amount            | No        | The maximum effective price, in the deposit asset, to pay for each LP Token received. |
| `LPTokenOut`  | [Currency Amount][] | Amount            | No        | How many of the AMM's LP Tokens to buy. |

### AMMDeposit Modes

> This transaction has several modes, depending on which flags you specify. Each mode expects a specific combination of fields. The modes fall into two categories: 

Reviewer 2: ```suggestion
This transaction has two deposit modes:
```

Author: The transaction has 5 modes, two of which are double-asset deposits and three of which are single-asset deposits. Each mode has a specific subset of allowed fields _and_ works differently in terms of how it decides how much money to deposit (where it places the limits). As much as I would like to say it has 2 modes, I think that ends up oversimplifying.

Reviewer 2: I misinterpreted the original meaning, then. Disregard.

> - **Double-asset deposits**, in which you provide both assets in the AMM's pool in proportions that match the balance of those assets already there. These deposits are not subject to a fee.

Reviewer 2: - **Double-asset deposits**, in which you provide both assets in the AMM's pool, proportional to the balance of the assets already there. These deposits are not subject to a fee.

> - **Single-asset deposits**, in which you provide only one of the AMM's two assets. You can envision a single-asset deposit as trading _some_ of the deposited asset for the other asset, then performing a double-asset deposit. The trading fee applies to the amount you would need to trade for, but not to the rest of the deposit; the amount of the fee is debited from the LP Tokens paid out. _For example, if the AMM's asset pool is split perfectly evenly between USD and EUR, and you try to deposit 100 USD, the amount of LP Tokens you receive is slightly less than if you had deposited 50 EUR + 50 USD, because you pay the trading fee to convert some of your USD to an equal amount of EUR._

Reviewer 2: The way this is worded is confusing. I don't think we should use the example of trading one asset for another, since it implies you're getting a withdrawal of the other asset. Maybe we elaborate on how the AMM will try to balance the difference, and in doing so subract trading fees from LP token payouts as it tries to balance the values.

Author: The way this _works_ is confusing; this wording is at least less confusing to me than anything else I've come up with. 

I'm not sure what you mean by elaborating on "how the AMM will try to balance the difference" instead. The AMM doesn't move anything around to try and make things balance, it just charges the trading fee on a percentage of your deposit. And it was helpful to me at least to know that it's calculated so that the result is identical to if you did a trade against the AMM first and then did a double-asset deposit.

Author: Changed to this:

> - **Single-asset deposits**, in which you provide only one of the AMM's two assets. The AMM charges a fee, debited from the LP Tokens paid out, based on how much your deposit shifts the balance of assets in the pool.

Then I moved some details to the Single Asset Deposit Fee Calculations section below:

> ### Single Asset Deposit Fee Calculations
>
> The fee for a single asset deposit is calculated to be the same as if you had used the AMM to trade part of the deposit amount for the other asset, then performed a double-asset deposit. The AMM's trading fee applies to the amount you would need to trade for, but not to the rest of the deposit. _For example, if the AMM's asset pool is split perfectly evenly between USD and EUR, and you try to deposit 100 USD, the amount of LP Tokens you receive is slightly less than if you had deposited 50 EUR + 50 USD, because you pay the trading fee to convert some of your USD to an equal amount of EUR._

The following combinations of fields indicate a **single asset deposit**:

| Flag Name           | Flag Value   | Fields Specified           | Meaning |
|---------------------|--------------|----------------------------|---------|
| `tfSingleAsset`     | `0x00080000` | `Amount` only              | Deposit exactly the specified amount of one asset, and receive an amount of LP Tokens based on the resulting share of the pool (minus fees). |
| `tfOneAssetLPToken` | `0x00200000` | `Amount` and `LPTokenOut`  | Deposit up to the specified amount of one asset, so that you receive exactly the specified amount of LP Tokens in return (after fees). |
| `tfLimitLPToken`    | `0x00400000` | `Amount` and `EPrice`      | Deposit up to the specified amount of one asset, but pay no more than the specified effective price per LP Token (after fees). |

Any other combination of these fields and flags is invalid.

### Single Asset Deposit Fee Calculations

> If you deposit one asset, the AMM charges a fee by reducing the total amount of LP Tokens it pays out, weighted by how much your deposit shifts the balance of assets in the pool. This is equivalent to if you had traded part of the deposit amount for the other asset, then performed a double-asset deposit. The formula for how many LP Tokens you receive for a double-asset deposit is as follows:

Reviewer 2: If we make the explanation for single asset deposits more concise in the previous section, we won't have to reiterate what it is here and can instead go straight into the fee calculation.

Author: Tweaked this section based on your feedback (see comment above for what I went with)
